File: src/main/java/com/waridley/ttv/logger/ChatLogger.java
package com.waridley.ttv.logger;

import com.github.twitch4j.chat.TwitchChat;
import com.github.twitch4j.chat.events.channel.*;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.List;

@Getter
@Slf4j
public class ChatLogger {
	
	protected final String channelName;
	protected String guestChannelName;
	protected final TwitchChat chat;
	protected List<MessageLogger> messageLoggers;
	
	public ChatLogger(TwitchChat chat, String channelName) {
		this.channelName = channelName;
		this.chat = chat;
		chat.joinChannel(channelName);
		log.info("Joined channel " + channelName);
		
		messageLoggers = new ArrayList<>();
		
		chat.getEventManager().onEvent(ChannelMessageEvent.class).subscribe(this::logMessage);
		chat.getEventManager().onEvent(HostOnEvent.class).subscribe(this::onHost);
		chat.getEventManager().onEvent(HostOffEvent.class).subscribe(this::onUnhost);
		chat.getEventManager().onEvent(ChannelJoinEvent.class).subscribe(this::userJoined);
		chat.getEventManager().onEvent(ChannelLeaveEvent.class).subscribe(this::userLeft);
		chat.getEventManager().onEvent(IRCMessageEvent.class).subscribe(this::onRawMsg);
	}
	
	public void addMessageLogger(MessageLogger logger) {
		messageLoggers.add(logger);
	}
	
	protected synchronized void userJoined(ChannelJoinEvent event) {
		log.debug("{} just joined {}'s chat!", event.getUser().getName(), event.getChannel().getName());
	}
	
	protected synchronized void userLeft(ChannelLeaveEvent event) {
		log.debug("{} just left {}'s chat... snowpoSOB", event.getUser().getName(), event.getChannel().getName());
	}
	
	protected synchronized void onHost(HostOnEvent event) {
		if(event.getChannel().getName().equalsIgnoreCase(channelName) && !event.getTargetChannel().getName().equalsIgnoreCase(guestChannelName)) {
			if(guestChannelName != null) chat.leaveChannel(guestChannelName);
			guestChannelName = event.getTargetChannel().getName();
			log.info("Now hosting {}", guestChannelName);
			chat.joinChannel(guestChannelName);
		}
	}
	
	protected synchronized void onUnhost(HostOffEvent event) {
		log.info("No longer hosting {}", event.getChannel().getName());
		chat.leaveChannel(event.getChannel().getName());
		if(event.getChannel().getName().equalsIgnoreCase(guestChannelName)) guestChannelName = null;
	}
	
	protected synchronized void onRawMsg(IRCMessageEvent event) {
		log.trace(event.toString());
	}
	
	protected void logMessage(ChannelMessageEvent event) {
		for(MessageLogger logger : messageLoggers) {
			logger.logMessage(event);
		}
	}
	
	@FunctionalInterface
	public interface  MessageLogger {
		void logMessage(ChannelMessageEvent event);
	}
}

File: src/main/java/com/waridley/ttv/logger/ChatLogger.kt
package com.waridley.ttv.logger

import com.github.twitch4j.chat.TwitchChat
import com.github.twitch4j.chat.events.channel.*
import org.slf4j.LoggerFactory
import java.util.*

open class ChatLogger(val chat: TwitchChat, val channelName: String) {
	var guestChannelName: String? = null
		protected set
	protected var messageLoggers: MutableList<MessageLogger>
	fun addMessageLogger(logger: MessageLogger) {
		messageLoggers.add(logger)
	}
	
	protected fun userJoined(event: ChannelJoinEvent) {
		log.debug("{} just joined {}'s chat!", event.user.name, event.channel.name)
	}
	
	protected fun userLeft(event: ChannelLeaveEvent) {
		log.debug("{} just left {}'s chat... snowpoSOB", event.user.name, event.channel.name)
	}
	
	protected fun onHost(event: HostOnEvent) {
		if (event.channel.name.equals(channelName, ignoreCase = true) && !event.targetChannel.name.equals(guestChannelName, ignoreCase = true)) {
			if (guestChannelName != null) chat.leaveChannel(guestChannelName)
			guestChannelName = event.targetChannel.name
			log.info("Now hosting {}", guestChannelName)
			chat.joinChannel(guestChannelName)
		}
	}
	
	protected fun onUnhost(event: HostOffEvent) {
		log.info("No longer hosting {}", event.channel.name)
		chat.leaveChannel(event.channel.name)
		if (event.channel.name.equals(guestChannelName, ignoreCase = true)) guestChannelName = null
	}
	
	protected fun onRawMsg(event: IRCMessageEvent) {
		log.trace(event.toString())
	}
	
	protected fun logMessage(event: ChannelMessageEvent) {
		for (logger in messageLoggers) {
			logger.logMessage(event)
		}
	}
	
	interface MessageLogger {
		fun logMessage(event: ChannelMessageEvent)
	}
	
	companion object {
		private val log = LoggerFactory.getLogger(ChatLogger::class.java)
	}
	
	init {
		chat.joinChannel(channelName)
		log.info("Joined channel $channelName")
		messageLoggers = ArrayList()
		chat.eventManager.onEvent(ChannelMessageEvent::class.java).subscribe { event: ChannelMessageEvent -> logMessage(event) }
		chat.eventManager.onEvent(HostOnEvent::class.java).subscribe { event: HostOnEvent -> onHost(event) }
		chat.eventManager.onEvent(HostOffEvent::class.java).subscribe { event: HostOffEvent -> onUnhost(event) }
		chat.eventManager.onEvent(ChannelJoinEvent::class.java).subscribe { event: ChannelJoinEvent -> userJoined(event) }
		chat.eventManager.onEvent(ChannelLeaveEvent::class.java).subscribe { event: ChannelLeaveEvent -> userLeft(event) }
		chat.eventManager.onEvent(IRCMessageEvent::class.java).subscribe { event: IRCMessageEvent -> onRawMsg(event) }
	}
}